lib/adf_diag.py

```diff
@@ -1017,66 +1028,135 @@ def derive_variables(self, vars_to_derive=None, ts_dir=None, overwrite=None):
-                        for v in ["FLNT", "FSNT"]
-                    ]
-                    constit_files = []
-                    for elem in input_files:
-                        constit_files += elem
-                else:
-                    ermsg = "FSNT and FLNT were not both present; RESTOM cannot be calculated."
-                    ermsg += " Please remove RESTOM from diag_var_list or find the relevant CAM files."
-                    raise FileNotFoundError(ermsg)
-                # create new file name for RESTOM
-                derived_file = constit_files[0].replace("FLNT", "RESTOM")
+            print(f"\t - deriving time series for {var}")
+
+            #Check whether there are parts to derive from and if there is an associated equation
+            vres = res.get(var, {})
+            if "derivable_from" in vres:
+                constit_list = vres['derivable_from']
+            else:
+                print("WARNING: No constituents listed in defaults config file, moving on")
+                continue
+
+            constit_files = []
+            for constit in constit_list:
+                if glob.glob(os.path.join(ts_dir, f"*.{constit}.*.nc")):
+                    constit_files.append(glob.glob(os.path.join(ts_dir, f"*.{constit}.*"))[0])
+
+            #Check if all the constituent files were found
+            if len(constit_files) != len(constit_list):
+                ermsg = f"Not all constituent files present; {var} cannot be calculated."
+                ermsg += f" Please remove {var} from diag_var_list or find the relevant CAM files."
+                print(ermsg)
+            else:
+                #Open a new dataset with all the constituent files/variables
+                ds = xr.open_mfdataset(constit_files)
+
+                # create new file name for derived variable
+                derived_file = constit_files[0].replace(constit_list[0], var)
+                print(derived_file,"\n")
                 if Path(derived_file).is_file():
                     if overwrite:
                         Path(derived_file).unlink()
                     else:
                         print(
-                            f"[{__name__}] Warning: RESTOM file was found and overwrite is False. Will use existing file."
+                            f"[{__name__}] Warning: '{var}' file was found and overwrite is False. Will use existing file."
                         )
-                        continue
-                # append FSNT to the file containing FLNT
-                os.system(f"ncks -A -v FLNT {constit_files[0]} {constit_files[1]}")
-                # create new file with the difference of FLNT and FSNT
-                os.system(
-                    f"ncap2 -s 'RESTOM=(FSNT-FLNT)' {constit_files[1]} {derived_file}"
-                )
+                        return None
+
+                variables = {}
+                for i in constit_list:
+                    variables[i] = ds[constit_list[0]].dims
```

Comment:
I'm a little confused about what this does. Looks like the keys of `variables` will be the names of the constituents of the derived quantity, and the values will all be the same tuple of the first constituent's dimensions.

`variables` then gets used below to loop through the dimensions.

I wonder if you get the same result if by replacing lines 1071-1092 with this:

```python
s = 0
for v in list_of_variables:
    s += ds[v]
s.name = var
```

and then either `ds[var] = s`, or just write `s.to_netcdf(...)` when it is ready for output (see my comment below about the output file).